### PR TITLE
UPSTREAM: 38112: Add json,yaml output format support to `oc create`, `oc apply`

### PR DESCRIPTION
**Pull request by juanvallejo**
Upstream: https://github.com/kubernetes/kubernetes/pull/38112
Related upstream issue: https://github.com/kubernetes/kubernetes/issues/37390

This patch adds the ability to specify an output format other than
"name" to `oc create ...`. It can be used in conjunction with the
`--dry-run` option.

Will add upstream PR

Due to the `mapper` being an [UnstructuredObject](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create.go#L114), using a `wide` format option results in the error:
```
error: unknown type &runtime.Unstructured{Object:map[string]interface {}{"status":map[string]interface {}{"lastVersion":1}, "kind":"BuildConfig", "apiVersion":"v1", "metadata":map[string]interface {}{"name":"gitauthtest", "namespace":"default", "selfLink":"/oapi/v1/namespaces/default/buildconfigs/gitauthtest", "uid":"39b2c5a1-b1b3-11e6-9a01-507b9dac96e1", "resourceVersion":"800167", "creationTimestamp":"2016-11-23T19:30:00Z", "labels":map[string]interface {}{"app":"gitauthtest", "template":"gitserver"}, "annotations":map[string]interface {}{"openshift.io/generated-by":"OpenShiftNewApp"}}, "spec":map[string]interface {}{"runPolicy":"Serial", "source":map[string]interface {}{"type":"Git", "git":map[string]interface {}{"uri":"http://gitserver-tokenauth.linux.xip.io/ruby-hello-world"}, "sourceSecret":map[string]interface {}{"name":"builder-token-nbme5"}}, "strategy":map[string]interface {}{"type":"Source", "sourceStrategy":map[string]interface {}{"from":map[string]interface {}{"kind":"ImageStreamTag", "namespace":"openshift", "name":"ruby:latest"}}}, "output":map[string]interface {}{}, "resources":map[string]interface {}{}, "postCommit":map[string]interface {}{}, "nodeSelector":interface {}(nil), "triggers":[]interface {}{}}}}
```

To mitigate this, an `unsupported output format` error message is printed when a format like this one is specified.

**Example**
```
$ oc create -f pkg/api/graph/test/bc-missing-output.yaml --dry-run -o yaml
apiVersion: v1
kind: BuildConfig
metadata:
  annotations:
    openshift.io/generated-by: OpenShiftNewApp
  creationTimestamp: 2015-11-18T14:41:17Z
  labels:
    app: gitauthtest
    template: gitserver
  name: gitauthtest
  namespace: default
spec:
  output: {}
  resources: {}
  source:
    git:
      uri: http://gitserver-tokenauth.linux.xip.io/ruby-hello-world
    sourceSecret:
      name: builder-token-nbme5
    type: Git
  strategy:
    sourceStrategy:
      from:
        kind: ImageStreamTag
        name: ruby:latest
        namespace: openshift
    type: Source
  triggers: []
status:
  lastVersion: 1
```

```
$ oc create -f pkg/api/graph/test/bc-missing-output.yaml --dry-run -o wide
error: That output format is not supoported
```

cc @openshift/cli-review

### CODE REVIEW

**Comment by fabianofranz** [comment]
[test]

**Comment by fabianofranz** [comment]
conformance flaked on https://github.com/openshift/origin/issues/11747 re[test]

**Comment by gilad207** [comment]
check

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/create.go [hunk 1]
```diff
@@ -150,7 +152,21 @@ func RunCreate(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *C
...
+			printer, generic, err := cmdutil.PrinterForCommand(cmd)
+			if err != nil {
+				return err
+			}
+			if !generic {
+				decodedObj, decoded := decodeUnstructuredObject(info.Object)
+				if !decoded {
+					return fmt.Errorf("Unsupported output format for the current object")
+				}
+				return f.PrintObject(cmd, registered.RESTMapper(), decodedObj, out)
```
Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/create.go]: @fabianofranz Added support for printing unstructured objects with a wide printer. PTAL
Sample output:

```
$ oc create -f pkg/api/graph/test/bc-missing-output.yaml --dry-run -o wide
NAME          TYPE      FROM      LATEST
gitauthtest   Source    Git       1
```

**Comment by juanvallejo** [comment]
networking check flaked on https://github.com/openshift/origin/issues/12091 re[test]

**Comment by juanvallejo** [comment]
@fabianofranz I can go ahead and open an Upstream PR if everything looks okay

**Comment by juanvallejo** [comment]
Related upstream PR https://github.com/kubernetes/kubernetes/pull/38112

**Comment by juanvallejo** [comment]
conformance check flaked on https://github.com/openshift/origin/issues/11662 re[test]

**Comment by juanvallejo** [comment]
check flaked on https://github.com/openshift/origin/issues/12157 re[test]

**Comment by juanvallejo** [comment]
check flaked on https://github.com/openshift/origin/issues/12157 re[test]

**Comment by juanvallejo** [comment]
integration flaked on https://github.com/openshift/origin/issues/8571
 re[test]

**Comment by juanvallejo** [comment]
integration flaked on https://github.com/openshift/origin/issues/8571
re[test]

**Comment by juanvallejo** [comment]
@fabianofranz [Upstream](https://github.com/kubernetes/kubernetes/pull/38112) merged!
This patch is up to date with upstream PR, however this PR also adds [this block to resource_printer, which handles unstructured objects](https://github.com/openshift/origin/pull/12015/files#diff-cefacfc976acfb112f45f3a46c736befR2247). It is missing from the current version of go we vendor, but [already exists upstream](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L2391).

**Comment by fabianofranz** [comment]
@juanvallejo cool, are you ok with tagging this as `post-rebase` and waiting until the next one?

**Comment by juanvallejo** [comment]
@fabianofranz definitely, it would remove the need to modify resource_printer if we wait until the rebase

**Comment by fabianofranz** [comment]
@juanvallejo actually this is just UPSTREAM commits. Are there any Origin issues or bugs this fixes? Otherwise I think it's just closing this one and waiting for the rebase.  :)

**Comment by juanvallejo** [comment]
@fabianofranz 
> @juanvallejo actually this is just UPSTREAM commits. Are there any Origin issues or bugs this fixes? 

Looks like we'll get all of the changes with the next rebase, closing this pr

**Pull request closed without merging**